### ISSUE TITLE
Link Mode

### ISSUE THREAD

**Issue by coolacid** [python-aprs/ax253#7]
Is there plans support Linked TNCs not just UI frames?

**Comment by masenf** [comment]
As you've probably discovered, nearly every ax25 parsing library says they "plan to support connected mode", but nobody ever gets around to it, because it actually has very complex semantics for how simple of a protocol it seems to be.

I don't see myself implementing non-UI frames in the foreseeable future, but I would welcome patches or collaboration if a motivated contributor wanted to help.

That said, my current solution for connected mode packet work is https://github.com/masenf/gensio-modems, which builds upon [gensio](https://github.com/cminyard/gensio) for which I provide python binaries on linux and macOS. It's not perfect, but I've been using this with Winlink and traditional BBS (both clients and servers) to great effect. It doesn't provide the raw ax25 packets, but instead exposes a socket interface over the ax25 transport.

**Comment by coolacid** [comment]
Neat tool. 

Yeah, connected mode has some nuance the further down the path it goes. 

I started to try and reply to a simple SABM packet, and stopped pretty quickly when I realised it would make a whole lot more sense to be part of the library. 

I may start building this part out in stages, wouldn't be fully complete, and not to the level of this library, but if I can get a working PoC that you could clean up and build upon - No guarantees, I have a bunch of dangling projects ;)

**Comment by masenf** [comment]
> I have a bunch of dangling projects ;)

100% know how this goes. I've got loads of projects / patches that scratch my itch, but for every 1 I publish, there's probably at least 3 that rot away in my homedir.

All that to say, if you get _anything_ going, I'll be happy to review your patch and try to get it in, even if it's not fully complete. If `ax253` can do connected mode packets at all, I think that alone will draw further contributors.

**Comment by coolacid** [comment]
Last night I was successful in getting a connection initiated from a "remote station". I've since forked and starting tracking my changes there. Took way to much digging then I should have done - it's been a while since I did any AX25 work, let alone in Python. 

Before I continue, I need to sit and think how this might be used. For my own functions, I'm looking to mimic LinBPQ which means for each connection I'll want to spawn a thread to an "application". Just not sure the data path should work between a main app, AX253 and the application thread. 

Two main options I see:

1: Let the main application handle mapping a Source/Dest address to a thread, and run the AX253 connection handler in the spawned thread so it only needs to handle connection state and it's timers. This becomes complicated spaghetti of call backs and classes. 

2: Register applications in the AX253 connection handler, and let it maintain states and threads. This might be easier to maintain, but puts a lot more weight on AX253. 

Where this gets even more complicated is when you have a "main" application that lets you choose other applications. IE: You might map SSID 10 to Winlink, but SSID 0 to a text based menu that can spawn a winlink connection.

**Comment by masenf** [comment]
Since ax253 is ultimately based on [asyncio](https://docs.python.org/3/library/asyncio.html), i think it will be beneficial to try to think within that framework of coroutines and awaitables.

In my mind, the ideal implementation looks like a "stream layer" `Protocol` that would sit on top of the [`FrameDecodeProtocol`](https://github.com/python-aprs/ax253/blob/67ba4681c7b52925622a0266f8cf4c0cc041e8fa/src/ax253/decode.py#L64) and provide connected mode semantics. So we could have the underlying `KISSProtocol` expose a `connect(source, dest)` async function returning something like a `AX25ConnectedModeReader` and  `AX25ConnectedModeWriter` for further interactions with the client application.

I don't really see a use case for extremely low level access to the connected mode protocol; do applications really want to manage timers and state? Or does an application just want the "stream metaphor" and try to be agnostic to the underlying transport? In that case, mimicing the interface of [`asyncio.open_connection`](https://docs.python.org/3/library/asyncio-stream.html#asyncio.open_connection) would allow applications to target _any_ transport that follows the asyncio paradigm (like raw TCP).

**Comment by coolacid** [comment]
I've never really worked with asyncio, and actively try and avoid it where ever possible for no other reason then I've never required it for any project. 

I don't forsee an application requiring low level access. 

What I might try and do is get a basic setup working, and let someone else take that and modify it into a better implementation with asyncio and example.

**Comment by masenf** [comment]
> I've never really worked with asyncio

This was also me before working on this project actually.

The main thing I like about asyncio is that nearly all IO (sockets and processes) in recent python3 releases uses it under the hood, so you can wait on multiple handles like with `select`, but without writing explicit loops and missing error cases.

It definitely takes some getting used to, and I always want to expose a blocking interface on top of the asyncio, since blocking calls are what most python programmers are accustomed to.

That said, yes, please proceed with the details of the implementation, timers, semantics, etc. I will help you wrap it in asyncio-ness if you like.

**Comment by coolacid** [comment]
Yeah, I suspect I'll bite the bullet and learn some asyncio - Given my goal of mimicking the LinBPQ switch, doing this in an event loop makes more sense then spawning a bunch of threads for different connections. 

I'm still going to first step this on getting the basics of having a simple incoming connection (outgoing can be a future step), and bidirectional `I` packets within the connection.
